src/js/index.tsx

```diff
@@ -206,8 +206,12 @@ initializeI18n((i18nError?: string) => {
         const user = selectUserState(state);
         // If user has no projects and is currently fetching projects, update state
         // to show spinner instead of empty projects-list.
-        if (user?.currently_fetching_repos && !projects.length) {
-          appStore.dispatch(projectsRefreshing());
+        if (user?.currently_fetching_repos) {
+          appStore.dispatch(
+            projects.length
+              ? { type: 'REFRESH_PROJECTS_REQUESTED' }
+              : projectsRefreshing(),
+          );
```

Comment:
@jgerigmeyer this seems to do what we want: disable the "Re-sync projects" button if the flag is stuck in the backend. Not sure if this was the right place to add it though

Reply:
@jerivas Practically, these two branches (`REFRESH_PROJECTS_REQUESTED` vs `projectsRefreshing()`) are equivalent. So the real change here is marking projects as "loading" even if there are already existing projects. The issue here is a race condition though, and this solves one problem but creates another.

When a user initially logs in, the front-end fetches the list of projects, and simultaneously the back-end [automatically triggers a `user.queue_refresh_repositories()`](https://github.com/oddbird/Metecho/blob/main/metecho/api/models.py#L1309-L1311). The problem is that this happens before the client has had a chance to subscribe to the relevant websocket channels, and often the job even _completes_ before the WS channel subscription is finalized. This means that the client usually does not receive the final `USER_REPOS_REFRESH` event, which would signal that the job is done and it's time to refetch the new list.

The prior "solution" to this (which was not really a solution) was to only show the loading indicator if there were no projects. _Usually_ this meant the following order of events:

1. User logs in.
2. Back-end triggers refresh.
3. Front-end loads user model, sees that the user is currently re-fetching projects.
4. Re-fetch repos job completes.
5. Front-end loads projects list.
6. Front-end subscribes to WS channels.

If (5) happens before (4) and the list was not empty, then the front end never knows that the job completed, and never re-fetches the updated list. If the list was empty, then the front end often hangs in the "loading" state until a page refresh (unless the WS subscription was able to happen before the job completed).

So this was already broken, in addition to the "what happens if the `user.currently_fetching_repos` flag is stuck on?" problem. But this change doesn't really fix it, but rather exposes the race condition issue to all users on every login (or almost every login, depending).

We need a better solution here, but I'm having trouble thinking of what it should be. If we can't come up with something reliable, maybe the front-end polls every second or so while `user.currently_fetching_repos` is on?

Reply:
@jgerigmeyer Interesting, didn't know so much was going on. Would it help to move this to the frontend explicitly requesting a project refresh? It seems we can get rid of some the uncertainty without that backend signal.

Reply:
> Would it help to move this to the frontend explicitly requesting the repos refresh? It seems we can get rid of some the uncertainty by removing that backend signal.

@jerivas Oh, that's an interesting proposal. It would add a bit of loading time on every login, but maybe that's okay. Or alternatively if we want to avoid that, we could skip the refresh unless the list is empty? I'll try it out and see how it feels.

Reply:
@jerivas In the end, explicitly requesting the refresh from the front-end only partially solved the issue, because there was still often a race condition where the WS subscription wasn't necessarily complete before the job finished. One option would be to wait until the WS subscription is made before even requesting the refresh or showing the app -- I played with that initially, but it felt a bit fragile to depend on WS before rendering.

Instead, I decided to just poll the API every second, to see if the job is done. If the job already completed before the WS subscription, then the front end will discover this after a 1 second delay. If the WS subscription was made before the job completes, then the WS event will come through normally and the polling will cancel. Hand-testing seems to confirm this approach is okay. What do you think about it?

https://github.com/oddbird/Metecho/commit/74dc92c86fd4762e04e3ac0a4cff01043efd0c72

Reply:
I think it's a bit complex, but I don't see any other way to comply with what we need, so LGTM 👍

I would suggest setting a limit to the retries so we don't DDOS ourselves if something goes wrong.

Reply:
Agreed, I'll add a limit. I also was hoping for a simpler solution, but in the end this felt like the "correct" and more robust approach -- if we're relying on WS events but don't want to wait until WS subscriptions complete (they can be slow), then I feel like we need a fallback like this.

Reply:
@jerivas Limit added in https://github.com/oddbird/Metecho/commit/4ad24fa06872571f3199474dd1ef4f4a6ba92253. Go ahead and merge if this PR looks okay to you?